go-controller/pkg/ovn/gateway.go

```diff
@@ -49,18 +53,42 @@ func (ovn *Controller) createGatewayVIPs(protocol kapi.Protocol, sourcePort int3
 			klog.Errorf("Gateway router %s does not have physical ip (%v)", gatewayRouter, err)
 			continue
 		}
+
+		vips := physicalIPs
+		if len(svcIPs) > 0 {
+			vips = svcIPs
+		}
```

Comment:
can you please add a comment explaining this? 
why and when do we use one or another vip?

Reply:
is the comment on the function good enough?
"// if empty svcIP is provided, then the physical IPs will be used for the node"

Reply:
why should we have an empty svcIP?

Reply:
its just how I utilize the function for 2 different purposes. At its core the function is used for adding LB entries to GR and per worker LBs. However, sometimes it is called to add a cluster IP service to the LBs, other times it is called to add the node port entry (svc IP would be empty in this case)

Reply:
Yep I agree it's hard to follow. On the calling site, you see this being called with nil and wonder if it's an error, here there is this not so clear if. I wonder if wrapping what's in the loop in a function and splitting this in two would make it more readable.

Something along the line of :

```suggestion
func (ovn *Controller) createPerNodeVIPsBalancers(svcIPs []string, protocol v1.Protocol, sourcePort int32, targetIPs []string, targetPort int32) error {
	klog.V(5).Infof("Creating Node VIPs - %s, %d, [%v], %d", protocol, sourcePort, targetIPs, targetPort)
	// Each gateway has a separate load-balancer for N/S traffic
	gatewayRouters, _, err := ovn.getOvnGateways()
	if err != nil {
		return err
	}

	for _, gatewayRouter := range gatewayRouters {
		ovn.createSingleNodeVIPs(gatewayRouter, svcIPs, protocol, sourcePort, targetIPs, targetPort)
	}
	return nil
}

func (ovn *Controller) createPerNodePhysicalIPBalancers(protocol v1.Protocol, sourcePort int32, targetIPs []string, targetPort int32) error {
	klog.V(5).Infof("Creating Node VIPs - %s, %d, [%v], %d", protocol, sourcePort, targetIPs, targetPort)
	// Each gateway has a separate load-balancer for N/S traffic
	gatewayRouters, _, err := gateway.GetOvnGateways()
	if err != nil {
		return err
	}

	for _, gatewayRouter := range gatewayRouters {
		physicalIPs, err := gateway.GetGatewayPhysicalIPs(gatewayRouter)
		if err != nil {
			klog.Errorf("Gateway router %s does not have physical ip (%v)", gatewayRouter, err)
			continue
		}
		ovn.createSingleNodeVIPs(gatewayRouter, physicalIPs, protocol, sourcePort, targetIPs, targetPort)
	}
	return nil
}

func (ovn *Controller) createSingleNodeVIPs(gwRouter string, vips []string, protocol v1.Protocol, sourcePort int32, targetIPs []string, targetPort int32) error {
	gatewayLB, err := gateway.GetGatewayLoadBalancer(gwRouter, protocol)
	if err != nil {
		klog.Errorf("Gateway router %s does not have load balancer (%v)",
			gwRouter, err)
		return nil
	}

	// If self ip is in target list, we need to use special IP to allow hairpin back to host
	newTargets := util.UpdateIPsSlice(targetIPs, vips, []string{types.V4HostMasqueradeIP, types.V6HostMasqueradeIP})

	// With the physical_ip:sourcePort as the VIP, add an entry in
	// 'load_balancer'.
	err = loadbalancer.CreateLoadBalancerVIPs(gatewayLB, vips, sourcePort, newTargets, targetPort)
	if err != nil {
		klog.Errorf("Failed to create VIP in load balancer %s - %v", gatewayLB, err)
		return nil
	}

	if config.Gateway.Mode == config.GatewayModeShared {
		workerNode := util.GetWorkerFromGatewayRouter(gwRouter)
		workerLB, err := loadbalancer.GetWorkerLoadBalancer(workerNode, protocol)
		if err != nil {
			klog.Errorf("Worker switch %s does not have load balancer (%v)", workerNode, err)
			return nil
		}
		err = loadbalancer.CreateLoadBalancerVIPs(workerLB, vips, sourcePort, targetIPs, targetPort)
		if err != nil {
			klog.Errorf("Failed to create VIP in load balancer %s - %v", workerLB, err)
			return nil
		}
	}
	return nil
}

```

Reply:
ok will incorporate this change

Reply:
just to clarify the suggested change is for the old controller functions, im only changing the new